### PR TITLE
Moved some of Inductor IR nodes to be frozen

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137410
* #137861
* __->__ #137859

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by Skylion007** [comment]
Heads up, while this isn't explicitly checked until Python 3.13 (after which it errors) any frozen class must have all it's subclasses also be frozen.

**Comment by Skylion007** [comment]
Also, if you are freezing all these classes too, you may want to consider `slots=True` later on as well for extra memory savings + performance

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -110,11 +110,11 @@
 from typing_extensions import dataclass_transform
 
 
-@dataclass_transform()
-def ir_dataclass(_cls):
+@dataclass_transform(frozen_default=True)
+def ir_dataclass(_cls=None, frozen: bool = False):
```
Comment (XuehaiPan) [inline torch/_inductor/ir.py:114]:
```suggestion
@dataclass_transform(frozen_default=True)
def ir_dataclass(cls=None, /, *, frozen: bool = True):
```

**Comment by Chillee** [comment]
@Skylion007
> Also, if you are freezing all these classes too, you may want to consider slots=True later on as well for extra memory savings + performance

I've found it useful to be able to set attributes in `__post_init__` still with `__setattr__`.

> Heads up, while this isn't explicitly checked until Python 3.13 (after which it errors) any frozen dataclass must have all it's subclasses also be frozen.

Are you referring to https://github.com/python/cpython/issues/109409? I think this is still checked in mypy

**Comment by Skylion007** [comment]
> I've found it useful to be able to set attributes in `__post_init__` still with `__setattr__`.
> 
> >

You can still set attributes in `__post_init__` with slots = True as long as their name and type is pre-declared?

**Comment by Skylion007** [comment]
> @Skylion007
> 
> > Also, if you are freezing all these classes too, you may want to consider slots=True later on as well for extra memory savings + performance
> 
> I've found it useful to be able to set attributes in `__post_init__` still with `__setattr__`.
> 
> > Heads up, while this isn't explicitly checked until Python 3.13 (after which it errors) any frozen dataclass must have all it's subclasses also be frozen.
> 
> Are you referring to [python/cpython#109409](https://github.com/python/cpython/issues/109409)? I think this is still checked in mypy

Cool, 3.13 explicitly checks it at runtime too

**Pull request closed without merging**